Assignment2Q2.java

```diff
@@ -0,0 +1,112 @@
+import java.util.LinkedList;
+
+/**
+ * Created by Alexandra on 23.05.2017.
+ */
+public class Assignment2Q2 {
+
+    public static class TreeNode <T> {
+        private TreeNode<T> left;
+        private TreeNode<T> right;
+        private T value;
+
+        public TreeNode(TreeNode<T> left, TreeNode<T> right, T value) {
+            if (value == null) {
+                throw new IllegalArgumentException();
+            }
+            this.left = left;
+            this.right = right;
+            this.value = value;
+        }
+
+        public TreeNode<T> getLeft() {
+            return left;
+        }
+
+        public TreeNode<T> getRight() {
+            return right;
+        }
+
+        public T getValue() {
+            return value;
+        }
+
+        public String toString() {
+            return value.toString() + '\t';
+        }
+
+        public T lowestCommonAncestor(T value1, T value2) {
+            // since there is only one path from a node to the root
+            // the lowest ancestor will be on this path
+            // and stack of ancestors will also include it
+            LinkedList<TreeNode<T>> stack = new LinkedList<>();
+            LinkedList<Integer> checkedChildren = new LinkedList<>();
+            // 0 - there is no any children equal to value1 or value2
+            // 1 - there is one child equal to value1 or value2
+            // 2 - there are both requested values as children
+            // so we are looking for the first node with mark 2
+            LinkedList<Integer> marks = new LinkedList<>();
+            stack.add(this);
+            checkedChildren.add(0);
+            marks.add(0);
+            // performing DFS
+            while (!stack.isEmpty()) {
+                TreeNode<T> top = stack.getLast();
+                int checked = checkedChildren.removeLast();
+                if (checked == 0) {
+                    checkedChildren.add(checked + 1);
+                    if (top.getLeft() != null) {
+                        // check the left subtree
+                        stack.add(top.getLeft());
+                        checkedChildren.add(0);
+                        marks.add(0);
+                    }
+                } else if (checked == 1) {
+                    checkedChildren.add(checked + 1);

```

Comment:
It is really hard to see from just looking at this code that the loop invariant that all 3 lists are the same length is maintained. If you don't restructure to a single list (which would be better) then It would be nice to always see 3 removes or 3 adds in a row, so it is easy to see the lists always stay the same length. The code as it is written is quite hard to follow.